src/frequenz/client/base/client.py

```diff
@@ -22,6 +22,95 @@ class BaseApiClient(abc.ABC, Generic[StubT, ChannelT]):
     This class provides a common interface for API clients that communicate with a API
     server. It is designed to be subclassed by specific API clients that provide a more
     specific interface.
+
+    Some extra tools are provided to make it easier to write API clients:
+
+    - [call_stub_method()][frequenz.client.base.client.call_stub_method] is a function
+        that calls a gRPC stub method and translates errors to API client errors.
+    - [GrpcStreamBroadcaster][frequenz.client.base.streaming.GrpcStreamBroadcaster] is
+        a class that helps sending messages from a gRPC stream to
+        a [Broadcast][frequenz.channels.Broadcast] channel.
+
+    Example:
+        This example illustrates how to create a simple API client that connects to a
+        gRPC server and calls a method on a stub.
+
+        ```python
+        from collections.abc import AsyncIterable
+        import grpc
+        from frequenz.client.base.client import BaseApiClient, call_stub_method
+        from frequenz.client.base.streaming import GrpcStreamBroadcaster
+        from frequenz.channels import Receiver
+
+        # These classes are normally generated by protoc
+        class ExampleRequest:
+            int_value: int
+            str_value: str
+
+        class ExampleResponse:
+            float_value: float
+
+        class ExampleStub:
+            async def example_method(
+                self,
+                request: ExampleRequest  # pylint: disable=unused-argument
+            ) -> ExampleResponse:
+                ...
+
+            def example_stream(self) -> AsyncIterable[ExampleResponse]:
+                ...
+        # End of generated classes
+
+        class ExampleResponseWrapper:
+            def __init__(self, response: ExampleResponse):
+                self.transformed_value = f"{response.float_value:.2f}"
+
+        class MyApiClient(BaseApiClient[ExampleStub, grpc.Channel]):
+            def __init__(self, server_url: str, *, connect: bool = True):
+                super().__init__(
+                    server_url, ExampleStub, grpc.Channel, connect=connect
+                )
+                self._broadcaster = GrpcStreamBroadcaster(
+                    "stream",
+                    lambda: self.stub.example_stream(ExampleRequest()),
+                    ExampleResponseWrapper,
+                )
+
+            async def example_method(
+                self, int_value: int, str_value: str
+            ) -> ExampleResponseWrapper:
+                return await call_stub_method(
+                    self,
+                    lambda: self.stub.example_method(
```

Comment:
I would suggest to use explicit parameter names here to avoid having *ahem* the reader accidentally thinking it is a parameter name :innocent:

Reply:
Haha, too much rust going on? :stuck_out_tongue: 

Will do.

Reply:
Looking at the code I'm not so convinced now. The function is called "call stub method", isn't it pretty obvious that what you are passing is... *ahem*, the **stub method**?

Reply:
I thinks forcing to use:

```py
call_stub_method(self, stub_method=lambda: self.stub.example_method())
```

Might be a bit too verbose.

Reply:
Maybe it should be:

```py
call_stub_method(self, this_is_really_the_stub_method=lambda: self.stub.example_method())
```

OK, :trollface: now, so I'll shut up :rofl: